File: app_watcher/src/main/java/com/tongchen/twatcher/gank/ui/fragment/CategoryFragment.java
package com.tongchen.twatcher.gank.ui.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;

import com.bumptech.glide.Glide;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;
import com.tongchen.twatcher.MainActivity;
import com.tongchen.twatcher.R;
import com.tongchen.twatcher.base.ui.fragment.BaseFragment;
import com.tongchen.twatcher.base.ui.fragment.MVPFragment;
import com.tongchen.twatcher.gank.model.entity.Category;
import com.tongchen.twatcher.gank.model.entity.GankResult;
import com.tongchen.twatcher.gank.model.entity.MultipleItem;
import com.tongchen.twatcher.gank.presenter.CategoryPresenter;
import com.tongchen.twatcher.gank.presenter.ICategoryPresenter;
import com.tongchen.twatcher.gank.ui.adapter.CategoryAdapter;
import com.tongchen.twatcher.gank.view.ICategoryView;
import com.tongchen.twatcher.util.LogUtils;
import com.tongchen.twatcher.widget.SimpleItemDecoration;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;


public class CategoryFragment extends MVPFragment<List<GankResult>, ICategoryView, ICategoryPresenter> implements ICategoryView,
        BaseQuickAdapter.OnItemClickListener, BaseQuickAdapter.RequestLoadMoreListener {

    private static final String ARG_CATEGORY = "category";

    private static final int SINGLE_SPAN_COUNT = 1;
    private static final int MULTIPLE_SPAN_COUNT = 2;

    public static final int TYPE_TEXT = 0;
    public static final int TYPE_IMG = 1;
    public static final int TYPE_MULTIPLE = 2;

    @Inject
    Context mContext;

    @BindView(R.id.smartRefreshLyt)
    SmartRefreshLayout mRefreshLayout;
    @BindView(R.id.recyclerlv_content)
    RecyclerView mContentRecyclerLv;

    private CategoryAdapter mContentAdapter;
    private List<GankResult> mData = new ArrayList<>();
    private List<MultipleItem> mMultipleItemList = new ArrayList<>();

    //  当前所处Tab的分类
    private Category mCategory;
    private String mRequestName;
    private int mPage = 1;
    private int mSpanCount = 1;
    private BaseFragment mContentFragment;

    public CategoryFragment() {

    }

    public static CategoryFragment newInstance(Category category) {
        CategoryFragment fragment = new CategoryFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_CATEGORY, category);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mCategory = getArguments().getParcelable(ARG_CATEGORY);
            mRequestName = mCategory.getRequestName();
        }
    }

    @Override
    protected void injectFragment() {
        getFragmentComponent().inject2Fragment(this);
    }

    @Override
    public int bindLayout() {
        return R.layout.gank_fragment_category;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (TextUtils.equals(mCategory.getCategoryName(), "福利")) {
            mSpanCount = MULTIPLE_SPAN_COUNT;
        } else {
            mSpanCount = SINGLE_SPAN_COUNT;
        }

        mRefreshLayout.setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                mPage = 1;
                mPresenter.getGankDataByPage(mRequestName, 10, mPage, CategoryPresenter.MODE_REFRESH);
            }
        });
        mRefreshLayout.setEnableLoadMore(false);

        GridLayoutManager linearLayoutManager = new GridLayoutManager(getActivity(), mSpanCount, LinearLayoutManager.VERTICAL, false);
        mContentRecyclerLv.setLayoutManager(linearLayoutManager);
        mContentAdapter = new CategoryAdapter(mMultipleItemList);
        mContentRecyclerLv.setAdapter(mContentAdapter);
        mContentRecyclerLv.addItemDecoration(new SimpleItemDecoration(mContext));
        mContentRecyclerLv.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                switch (newState) {
                    case RecyclerView.SCROLL_STATE_IDLE:
                        //  滑动停止后再加载图片
                        Glide.with(mContext).resumeRequests();
                        break;
                    case RecyclerView.SCROLL_STATE_DRAGGING:
                    case RecyclerView.SCROLL_STATE_SETTLING:
                        //  滑动时暂停加载图片
                        Glide.with(mContext).pauseRequests();
                        break;
                }
            }
        });

        mContentAdapter.openLoadAnimation(BaseQuickAdapter.SCALEIN);
        mContentAdapter.disableLoadMoreIfNotFullPage(mContentRecyclerLv);
        //  设置上拉加载更多
        mContentAdapter.setOnLoadMoreListener(this, mContentRecyclerLv);
        mContentAdapter.setOnItemClickListener(this);

        mPresenter.getGankDataByPage(mRequestName, 10, mPage, CategoryPresenter.MODE_REFRESH);
    }

    @Override
    public void onLoadMoreRequested() {
        mPresenter.getGankDataByPage(mRequestName, 10, ++mPage, CategoryPresenter.MODE_MORE);
    }

    @Override
    public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
        MultipleItem item = (MultipleItem) adapter.getData().get(position);
        if (mActivity instanceof MainActivity) {
            if (TextUtils.equals(item.getData().getType(), "福利")) {
                mContentFragment = ContentPicFragment.newInstance(item.getData());
            } else {
                mContentFragment = ContentTextFragment.newInstance(item.getData());
            }
            ((MainActivity) mActivity).startFragment(mContentFragment);
        }
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void refreshSucceed(List<GankResult> result) {
        LogUtils.d("CategoryFragment", "refreshSucceed---" + mCategory);
        mRefreshLayout.finishRefresh();

        mData.clear();
        mData.addAll(result);
        mMultipleItemList.clear();

        for (GankResult data : mData) {
            if (TextUtils.equals(data.getType(), "福利")) {
                mMultipleItemList.add(new MultipleItem(MultipleItem.TYPE_IMG, data));
            } else {
                mMultipleItemList.add(new MultipleItem(MultipleItem.TYPE_TEXT, data));
            }
        }
        mContentAdapter.notifyDataSetChanged();
    }

    @Override
    public void refreshFailed(String errorMsg) {
        mRefreshLayout.finishRefresh();

        LogUtils.d("CategoryFragment", "refreshFailed---" + errorMsg);
    }

    @Override
    public void loadMoreSucceed(List<GankResult> result) {
        LogUtils.d("CategoryFragment", "loadMoreSucceed---" + result.toString());
        removeIncorrectData(result);
        mData.addAll(result);
        mMultipleItemList.clear();

        for (GankResult data : mData) {
            if (TextUtils.equals(data.getType(), "福利")) {
                mMultipleItemList.add(new MultipleItem(MultipleItem.TYPE_IMG, data));
            } else {
                mMultipleItemList.add(new MultipleItem(MultipleItem.TYPE_TEXT, data));
            }
        }
        mContentAdapter.notifyDataSetChanged();
        mContentAdapter.loadMoreComplete();
    }

    @Override
    public void loadMoreFailed(String errorMsg) {
        LogUtils.d("CategoryFragment", "loadMoreFailed---" + errorMsg);
        mContentAdapter.loadMoreFail();
    }

    //  移除有问题的数据
    private void removeIncorrectData(List<GankResult> results) {
        List<GankResult> incorrectDatas = new ArrayList<>();
        for (int i = 0; i < results.size(); i++) {
            GankResult result = results.get(i);
            if (TextUtils.equals(result.getType(), "福利")) {
                //  url 不以jpg、jpeg结尾或者包含 7xi8d6.com（此网址挂了）或者来自的 img.gank.io（证书无效）的移除
                if (!(result.getUrl().endsWith(".jpg") || result.getUrl().endsWith(".jpeg"))
                        || result.getUrl().contains("7xi8d6.com")
                        || result.getUrl().contains("img.gank.io")) {
                    incorrectDatas.add(result);
                }
            }
        }
        results.removeAll(incorrectDatas);
    }

    //  返回到顶部
    public void back2Top() {
        mContentRecyclerLv.smoothScrollToPosition(0);
    }
}

File: app_watcher/src/main/java/com/tongchen/twatcher/gank/ui/fragment/CategoryFragment.kt
package com.tongchen.twatcher.gank.ui.fragment

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import com.bumptech.glide.Glide
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener
import com.tongchen.twatcher.MainActivity
import com.tongchen.twatcher.R
import com.tongchen.twatcher.base.ui.fragment.BaseFragment
import com.tongchen.twatcher.base.ui.fragment.MVPFragment
import com.tongchen.twatcher.gank.model.entity.Category
import com.tongchen.twatcher.gank.model.entity.GankResult
import com.tongchen.twatcher.gank.presenter.CategoryPresenter
import com.tongchen.twatcher.gank.presenter.ICategoryPresenter
import com.tongchen.twatcher.base.ui.BaseRecyclerAdapter
import com.tongchen.twatcher.gank.ui.adapter.CategoryAdapter
import com.tongchen.twatcher.gank.view.ICategoryView
import com.tongchen.twatcher.util.LogUtils
import com.tongchen.twatcher.widget.SimpleItemDecoration
import kotlinx.android.synthetic.main.gank_fragment_category.*
import javax.inject.Inject

/**
 * @author TongChen
 * @date 2019/11/20  15:32
 * <p>
 * Desc:
 */
class CategoryFragment : MVPFragment<MutableList<GankResult>, ICategoryView, ICategoryPresenter>(), ICategoryView {

    private lateinit var mContentAdapter: CategoryAdapter
    private val mData = mutableListOf<GankResult>()

    //  当前所处Tab的分类
    private var mCategory: Category? = null
    private var mRequestName: String? = null
    private var mPage = 1
    private var mSpanCount = 1
    protected var mContentFragment: BaseFragment? = null

    @Inject
    lateinit var mContext: Context

    companion object {
        internal val SINGLE_SPAN_COUNT = 1
        internal val MULTIPLE_SPAN_COUNT = 2

        private val ARG_CATEGORY = "category"

        fun newInstance(category: Category): CategoryFragment {
            val fragment = CategoryFragment()
            val bundle = Bundle()
            bundle.putParcelable(ARG_CATEGORY, category)
            fragment.arguments = bundle
            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mCategory = arguments?.getParcelable(ARG_CATEGORY)
            mRequestName = mCategory?.requestName
        }
    }

    override fun injectFragment() {
        fragmentComponent.inject2Fragment(this)
    }

    override fun bindLayout() = R.layout.gank_fragment_category

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mSpanCount = if (mRequestName == "福利") {

            MULTIPLE_SPAN_COUNT
        } else {
            SINGLE_SPAN_COUNT
        }

        smartRefreshLyt.setOnRefreshLoadMoreListener(object : OnRefreshLoadMoreListener {
            override fun onRefresh(refreshLayout: RefreshLayout) {
                mPage = 1
                mPresenter.getGankDataByPage(mRequestName, 10, mPage, CategoryPresenter.MODE_REFRESH)
            }

            override fun onLoadMore(refreshLayout: RefreshLayout) {
                mPresenter.getGankDataByPage(mRequestName, 10, ++mPage, CategoryPresenter.MODE_MORE)
            }
        })

        val layoutManager = GridLayoutManager(mContext, mSpanCount, GridLayoutManager.VERTICAL, false)
        recyclerlv_content.layoutManager = layoutManager
        mContentAdapter = CategoryAdapter(mContext, mSpanCount, mData)
        recyclerlv_content.adapter = mContentAdapter
        recyclerlv_content.addItemDecoration(SimpleItemDecoration(mContext))
        recyclerlv_content.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                when (newState) {
                    //  滑动停止后再加载图片
                    RecyclerView.SCROLL_STATE_IDLE -> Glide.with(mContext).resumeRequests()
                    //  滑动时暂停加载图片
                    RecyclerView.SCROLL_STATE_DRAGGING,
                    RecyclerView.SCROLL_STATE_SETTLING -> Glide.with(mContext).pauseRequests()
                }
            }
        })

        mContentAdapter.setOnItemClickListener(object : BaseRecyclerAdapter.OnItemClickListener {
            override fun onItemClick(itemView: View, position: Int) {
                val itemData = mData[position]
                if (mActivity is MainActivity) {
                    if (itemData.type == "福利") {
                        mContentFragment = ContentPicFragment.newInstance(itemData)

                    } else {
                        mContentFragment = ContentTextFragment.newInstance(itemData)
                    }
                    (mActivity as MainActivity).startFragment(mContentFragment)
                }
            }
        })

        mPresenter.getGankDataByPage(mRequestName, 10, mPage, CategoryPresenter.MODE_REFRESH)
    }

    override fun showLoading() {

    }

    override fun hideLoading() {

    }

    override fun refreshSucceed(result: MutableList<GankResult>?) {
        smartRefreshLyt.finishRefresh(true)

        mData.clear()
        if (result != null) {
            mData.addAll(result)
            mContentAdapter.notifyDataSetChanged()
        }

        LogUtils.d("CategoryFragment", "refreshSucceed---" + mCategory + ", size:${result?.size}")
    }

    override fun refreshFailed(errorMsg: String?) {
        smartRefreshLyt.finishRefresh(false)

        LogUtils.d("CategoryFragment", "refreshFailed---$errorMsg")
    }

    override fun loadMoreSucceed(result: MutableList<GankResult>?) {
        smartRefreshLyt.finishLoadMore(true)

        if (result != null) {
            removeIncorrectData(result)
            mData.addAll(result)
            mContentAdapter.notifyDataSetChanged()
        }
        LogUtils.d("CategoryFragment", "loadMoreSucceed---${result.toString()}")
    }

    override fun loadMoreFailed(errorMsg: String?) {
        smartRefreshLyt.finishLoadMore(false)

        LogUtils.d("CategoryFragment", "loadMoreFailed---$errorMsg")
    }

    private fun removeIncorrectData(result: MutableList<GankResult>) {
        val incorrectData = mutableListOf<GankResult>()
        for (element in result) {
            if (element.type == "福利") {
                //  url 不以jpg、jpeg结尾或者包含 7xi8d6.com（此网址挂了）或者来自的 img.gank.io（证书无效）的移除
                if (!(element.url.endsWith(".jpg") || element.url.endsWith(".jpeg"))
                        || element.url.contains("7xi8d6.com")
                        || element.url.contains("img.gank.io")) {
                    incorrectData.add(element)
                }
            }
        }
        result.removeAll(incorrectData)
    }

    //  返回到顶部
    fun back2Top() {
        recyclerlv_content.smoothScrollToPosition(0)
    }
}